Source/Parallelization/WarpXComm.cpp

```diff
@@ -171,136 +171,191 @@ WarpX::UpdateAuxilaryDataStagToNodal ()
-                    warpx_interp(j, k, l, by_aux, by_fp, by_cp, by_c, By_fp_stag, By_cp_stag, refinement_ratio);
-                    warpx_interp(j, k, l, bz_aux, bz_fp, bz_cp, bz_c, Bz_fp_stag, Bz_cp_stag, refinement_ratio);
-                });
+                    Array4<Real> const& bx_aux = Bfield_aux[lev][0]->array(mfi);
+                    Array4<Real> const& by_aux = Bfield_aux[lev][1]->array(mfi);
+                    Array4<Real> const& bz_aux = Bfield_aux[lev][2]->array(mfi);
+                    Array4<Real const> const& bx_fp = Bfield_fp[lev][0]->const_array(mfi);
+                    Array4<Real const> const& by_fp = Bfield_fp[lev][1]->const_array(mfi);
+                    Array4<Real const> const& bz_fp = Bfield_fp[lev][2]->const_array(mfi);
+                    Array4<Real const> const& bx_cp = Bfield_cp[lev][0]->const_array(mfi);
+                    Array4<Real const> const& by_cp = Bfield_cp[lev][1]->const_array(mfi);
+                    Array4<Real const> const& bz_cp = Bfield_cp[lev][2]->const_array(mfi);
+                    Array4<Real const> const& bx_c = Btmp[0]->const_array(mfi);
+                    Array4<Real const> const& by_c = Btmp[1]->const_array(mfi);
+                    Array4<Real const> const& bz_c = Btmp[2]->const_array(mfi);
+
+                    const Box& bx = mfi.growntilebox();
+                    amrex::ParallelFor(bx,
+                    [=] AMREX_GPU_DEVICE (int j, int k, int l) noexcept
+                    {
+                        warpx_interp(j, k, l, bx_aux, bx_fp, bx_cp, bx_c, Bx_fp_stag, Bx_cp_stag, refinement_ratio);
+                        warpx_interp(j, k, l, by_aux, by_fp, by_cp, by_c, By_fp_stag, By_cp_stag, refinement_ratio);
+                        warpx_interp(j, k, l, bz_aux, bz_fp, bz_cp, bz_c, Bz_fp_stag, Bz_cp_stag, refinement_ratio);
+                    });
+                }
+            }
+            else {
+                const amrex::IntVect& Bx_fp_stag = Bfield_fp[lev][0]->ixType().toIntVect();
+                const amrex::IntVect& By_fp_stag = Bfield_fp[lev][1]->ixType().toIntVect();
+                const amrex::IntVect& Bz_fp_stag = Bfield_fp[lev][2]->ixType().toIntVect();
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+                for (MFIter mfi(*Bfield_aux[lev][0], TilingIfNotGPU()); mfi.isValid(); ++mfi)
+                {
+                    Array4<Real> const& bx_aux = Bfield_aux[lev][0]->array(mfi);
+                    Array4<Real> const& by_aux = Bfield_aux[lev][1]->array(mfi);
+                    Array4<Real> const& bz_aux = Bfield_aux[lev][2]->array(mfi);
+                    Array4<Real const> const& bx_fp = Bfield_fp[lev][0]->const_array(mfi);
+                    Array4<Real const> const& by_fp = Bfield_fp[lev][1]->const_array(mfi);
+                    Array4<Real const> const& bz_fp = Bfield_fp[lev][2]->const_array(mfi);
+
+                    const Box& bx = mfi.growntilebox();
```

Comment:
@WeiqunZhang 
Another question for you. It seems like here, for the B field, we use `growntilebox`, while below, for the E field, line 350, we use `fabbox`. I would expect that there is no reason for using different functions for B and E. Which one do you think is the most correct here, and why?

Reply:
I think it's better to use `growntilebox` and add `TilingIfNotGPU()` for the E field loop. For GPU, the two loops are the same. But for CPU, the tiling version should be better due to cache blocking and more OpenMP parallelism.